### PR TITLE
added single table inheritance to Post model

### PR DESCRIPTION
**Pull request by mattKendon**
Firstly, great library!

I ran into a number of issues where I wanted to override behaviour in some of my custom post types, but I couldn't reliably do this as the objects that I got back did not necessarily match the classes that I wanted them to be.

The most obvious type of behaviour to be overwritten would be permalink for the post. Pages would need to have one permalink structure, while a custom post type would need another. Adding a permalink attribute to the models that I could override depending on the post type wouldn't be feasible unless I could always know that the object would always be of the correct class.

Thankfully the wp_posts table lends itself to the Single Table Inheritance pattern perfectly, and after doing a little bit of research I was able to easily implement it with only a few lines of code. 

Now this means that however you get your posts out, they will always be of the correct class so that behaviours can be reliably overwritten.

I've added a new static method on the Post class that allows you to register a post type and associated class. I also overwrote the behaviour of the newFromModel() method so that it now fetches the correct class from the array of registered post types.

I have added two tests to cover this behaviour (one for registering, and one for clearing the registered posts as might need to happen).

I have also added a section to the README file to explain this new behaviour.

### CODE REVIEW

**Comment by mattKendon** [comment]
Please ignore my other pull request. I stupidly made the pull request to your dev branch, but my work was based off of master, which would have led to some unnecessary pollution of your git repository.

**Comment by jgrossi** [comment]
Hi @mattKendon! First of all thanks for the contribution. Let me understand what you did. The goal of this behavior is to return `Video` objects instead of `Post`, am I right?

There is the `$postType` attribute, that allow you to create the `Video` and set `$postType = 'video'`, so you just do `Video::all()` for example and get instances of `Video` classes, not `Post`. If you use `Post::type('video')-get()` you have `Post` objects.

What do you think about? And thanks for the contribution again.

File: src/Post.php [hunk 1]
```diff
@@ -526,5 +529,50 @@ public function setUpdatedAt($value)
         return parent::setUpdatedAt($value); // TODO: Change the autogenerated stub
     }
 
+    public function newFromBuilder($attributes = [], $connection = null)
```
Comment (josec89) [inline src/Post.php]: Could you add DocBlock here?

**Comment by jgrossi** [comment]
@mattKendon about the docblock, what to want to be there?

**Comment by mattKendon** [comment]
@jgrossi the goal of this behaviour is more specifically to always return `Video` objects when fetching posts that have a `post_type` of video.

The `Video::all()` works fine as long as you are only fetching posts with a `post_type` of video, however there are many use-cases for getting out a number of posts based on their IDs and not knowing beforehand what type of posts they are. 

I do this all the time with meta fields, where I might allow a user to connect one post with another (perhaps using the relationship field from the ACF plugin). The ID of the connected post is saved to the `wp_postmeta` table, but when I fetch it I cannot know beforehand what type of post it is, but I still want it to be a `Video` object if that is the type of post it is.

**Comment by mattKendon** [comment]
I've add the docbock to to the newFromBuilder method, so hopefully that will explain what I'm overriding a little better.

**Comment by jgrossi** [comment]
So, when using `->type('video')` it will always return the `Video` post instance?

**Comment by mattKendon** [comment]
Yes. And when using `::find(5)` where `5` is the ID for a video post, it will always return a `Video` instance.

**Comment by jgrossi** [comment]
Very nice @mattKendon! This work to scopes too? Like `where` or even custom scopes?

**Comment by mattKendon** [comment]
Yes, as they all work through the Builder, and this method is the last part of that chain. However, just to be safe I'll add a call to a scope method to make sure that it works.

**Comment by mattKendon** [comment]
Ignore that last comment. I use the `type` method in the test already so that is covered.

**Comment by jgrossi** [comment]
Very nice job indeed :-)

Can you create the pull request to the `dev` branch? Make sure to follow the instructions in the `README.md` and just send the PR to us! 

Thanks for the job :-) and welcome!

**Comment by mattKendon** [comment]
Created new pull request onto the dev branch

File: tests/PostTest.php [hunk 2]
```diff
@@ -149,4 +149,23 @@ public function testCustomFieldWithAccessors()
         $this->assertEquals($post->title, 'Hello world!');
         $this->assertEquals($post->meta->title, 'New title');
     }
+
+    public function testSingleTableInheritance()
```
Comment (josec89) [inline tests/PostTest.php]: Could you add DocBlock to these methods also :D?

**Pull request closed without merging**